### PR TITLE
Add Comprehensions of Lists, Sets and Maps and generator expressions

### LINKED ISSUE #1714: GeneratorExp

### LINKED ISSUE #1715: ListComp

### LINKED ISSUE #1716: SetComp

### LINKED ISSUE #1717: DictComp

### PR DESCRIPTION
**Pull request by KuechA**
This PR aims to enable list comprehensions and similar constructs.

Tasks:
* [x] Add a node which will hold the comprehension
* [x] Add translation for `Python.AST.ListComp`
* [x] Add translation for `Python.AST.SetComp`
* [x] Add translation for `Python.AST.DictComp`
* [x] Add translation for `Python.AST.GeneratorExp`
* [x] Connect the EOG of the new nodes
* [x] Connect the DFG of the new nodes in `DFGPass`
* [x] Handle the new nodes in the `ControlFlowSensitiveDFGPass`
* [x] Add some tests

Fixes #1714 
Fixes #1715 
Fixes #1716 
Fixes #1717

### CODE REVIEW

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/CollectionComprehension.kt [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+         * variable declaration or a reference (probably to a new variable).
+         */
+        var variable by unwrapping(ComprehensionExpression::variableEdge)
+
+        @Relationship("ITERABLE") var iterableEdge = astOptionalEdgeOf<Statement>()
+
+        /** This field contains the iteration subject of the loop. */
+        var iterable by unwrapping(ComprehensionExpression::iterableEdge)
+
+        @Relationship("PREDICATE") var predicateEdges = astEdgesOf<Statement>()
```
Comment (KuechA) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/CollectionComprehension.kt:72]: Single statement instead of list.

**Comment by KuechA** [comment]
@konradweiss please check and update the EOG implementation and specs

**Review by konradweiss — APPROVED** [review]
LGTM

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/ComprehensionExpression.kt [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+package de.fraunhofer.aisec.cpg.graph.statements.expressions
+
+import de.fraunhofer.aisec.cpg.graph.AccessValues
+import de.fraunhofer.aisec.cpg.graph.ArgumentHolder
+import de.fraunhofer.aisec.cpg.graph.edges.ast.astEdgeOf
+import de.fraunhofer.aisec.cpg.graph.edges.ast.astOptionalEdgeOf
+import de.fraunhofer.aisec.cpg.graph.edges.unwrapping
+import de.fraunhofer.aisec.cpg.graph.statements.Statement
+import java.util.Objects
+import kotlin.collections.plusAssign
```
Comment (maximiliankaul) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/ComprehensionExpression.kt:35]:
```suggestion
```

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/ComprehensionExpression.kt [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        if (other !is ComprehensionExpression) return false
+        return super.equals(other) &&
+            variable == other.variable &&
+            iterable == other.iterable &&
+            predicate == other.predicate
+    }
+
+    override fun hashCode() = Objects.hash(super.hashCode(), variable, iterable, predicate)
+
+    override fun addArgument(expression: Expression) {
```
Comment (maximiliankaul) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/ComprehensionExpression.kt:93]: I don't get the need for this function.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/ComprehensionExpression.kt [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        if (this.variable is ProblemExpression) {
+            this.variable = expression
+        } else if (this.iterable is ProblemExpression) {
+            this.iterable = expression
+        } else {
+            this.predicate = expression
+        }
+    }
+
+    override fun replaceArgument(old: Expression, new: Expression): Boolean {
```
Comment (maximiliankaul) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/ComprehensionExpression.kt:103]: I don't get the need for this function.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/ComprehensionExpression.kt [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        }
+
+        if (this.predicate == old) {
+            this.predicate = new
+            return true
+        }
+        return false
+    }
+
+    override fun hasArgument(expression: Expression): Boolean {
```
Comment (maximiliankaul) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/ComprehensionExpression.kt:121]: I don't get the need for this function.

**Review by maximiliankaul — APPROVED** [review]

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/ComprehensionExpression.kt (hunk 5 above)
Comment (KuechA) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/ComprehensionExpression.kt:121]: Required by implementing ArgumentHolder. Adds stuff for Fluent DSL.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/ComprehensionExpression.kt (hunk 4 above)
Comment (KuechA) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/ComprehensionExpression.kt:103]: Required by implementing ArgumentHolder. Adds stuff for Fluent DSL.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/ComprehensionExpression.kt (hunk 3 above)
Comment (KuechA) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/ComprehensionExpression.kt:93]: Required by implementing ArgumentHolder. Adds stuff for Fluent DSL.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/builder/Fluent.kt [hunk 6]
```diff
@@ -325,6 +326,47 @@ fun LanguageFrontend<*, *>.subscriptExpr(
...
+): ComprehensionExpression {
+    val node = newComprehensionExpression()
+
+    if (init != null) {
+        init(node)
+    }
+
+    // Only add this to an argument holder if the nearest holder is an argument holder
+    val holder = this@Holder
+    if (holder is ArgumentHolder) {
```
[the remark this answers is not in the document]
Comment (KuechA) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/builder/Fluent.kt:363]: We want to add it to the collection comprehension which is an ArgumentHolder. So I think we can be sure about this here.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/builder/Fluent.kt [hunk 7]
```diff
@@ -325,6 +326,47 @@ fun LanguageFrontend<*, *>.subscriptExpr(
...
+        init(node)
+    }
+
+    // Only add this to an argument holder if the nearest holder is an argument holder
+    val holder = this@Holder
+    if (holder is StatementHolder) {
+        holder += node
+    }
+    if (holder is ArgumentHolder) {
+        holder += node
```
[the remark this answers is not in the document]
Comment (KuechA) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/builder/Fluent.kt:345]: Added an else

File: cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/graph/FluentTest.kt [hunk 8]
```diff
@@ -179,4 +184,184 @@ class FluentTest {
...
+                                            listComp {
+                                                ref("i")
+                                                compExpr {
+                                                    ref("i")
+                                                    ref("someIterable")
+                                                    ref("i") gt
+                                                        literal(
+                                                            5,
+                                                            t("int")
+                                                        ) // TODO: This line doesn't work as
```
[the remark this answers is not in the document]
Comment (KuechA) [reply cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/graph/FluentTest.kt:214]: Some nodes were added multiple times to different parent nodes. No idea why but I can't replicate right now. Will be fixed with removing FluentDSL

File: cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/graph/FluentTest.kt [hunk 9]
```diff
@@ -179,4 +184,184 @@ class FluentTest {
...
+                                            listComp {
+                                                ref("i")
+                                                compExpr {
+                                                    this.variable = declare { variable("i") }
+                                                    ref("someIterable")
+                                                    ref("i") gt
+                                                        literal(
+                                                            5,
+                                                            t("int")
+                                                        ) // TODO: This line doesn't work as
```
[the remark this answers is not in the document]
Comment (KuechA) [reply cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/graph/FluentTest.kt:272]: Some nodes were added multiple times to different parent nodes. No idea why but I can't replicate right now. Will be fixed by removing FluentDSL

File: cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/graph/FluentTest.kt [hunk 10]
```diff
@@ -179,4 +184,184 @@ class FluentTest {
...
+        assertEquals(1, listComp.comprehensionExpressions.size)
+        val compExpr = listComp.comprehensionExpressions.single()
+        assertIs<ComprehensionExpression>(compExpr)
+        val variableDecl = compExpr.variable
+        assertIs<DeclarationStatement>(variableDecl)
+        assertLocalName("i", variableDecl.declarations[0])
+        assertLocalName("y", variableDecl.declarations[1])
+        assertIs<Reference>(compExpr.iterable)
+        assertLocalName("someIterable", compExpr.iterable)
+        // assertEquals(1, compExpr.predicates.size)
```
[the remark this answers is not in the document]
Comment (KuechA) [reply cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/graph/FluentTest.kt:365]: No longer relevant. Changed

File: cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/graph/FluentTest.kt [hunk 11]
```diff
@@ -179,4 +184,184 @@ class FluentTest {
...
+        assertIs<Reference>(listComp.statement)
+        assertLocalName("i", listComp.statement)
+        assertEquals(1, listComp.comprehensionExpressions.size)
+        val compExpr = listComp.comprehensionExpressions.single()
+        assertIs<ComprehensionExpression>(compExpr)
+        assertIs<Reference>(compExpr.variable)
+        assertLocalName("i", compExpr.variable)
+        assertIs<Reference>(compExpr.iterable)
+        assertLocalName("someIterable", compExpr.iterable)
+        // assertEquals(1, compExpr.predicates.size)
```
[the remark this answers is not in the document]
Comment (KuechA) [reply cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/graph/FluentTest.kt:243]: No longer relevant. Changed

File: cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/graph/FluentTest.kt [hunk 12]
```diff
@@ -179,4 +184,184 @@ class FluentTest {
...
+        assertLocalName("i", listComp.statement)
+        assertEquals(1, listComp.comprehensionExpressions.size)
+        val compExpr = listComp.comprehensionExpressions.single()
+        assertIs<ComprehensionExpression>(compExpr)
+        val variableDecl = compExpr.variable
+        assertIs<DeclarationStatement>(variableDecl)
+        assertLocalName("i", variableDecl.singleDeclaration)
+        assertIs<Reference>(compExpr.iterable)
+        assertLocalName("someIterable", compExpr.iterable)
+        // assertEquals(1, compExpr.predicates.size)
```
[the remark this answers is not in the document]
Comment (KuechA) [reply cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/graph/FluentTest.kt:302]: No longer relevant. Changed

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt [hunk 13]
```diff
@@ -82,6 +83,73 @@ class ExpressionHandler(frontend: PythonLanguageFrontend) :
...
+    /**
+     * Translates a Python
+     * [`comprehension`](https://docs.python.org/3/library/ast.html#ast.comprehension) into a
+     * [ComprehensionExpression].
+     *
+     * Connects multiple predicates by "AND".
+     */
+    private fun handleComprehension(node: Python.AST.comprehension): ComprehensionExpression {
+        return newComprehensionExpression(node).apply {
+            this.variable = handle(node.target)
```
[the remark this answers is not in the document]
Comment (KuechA) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt:95]: you can get rid of it but it doesn't harm to write it down

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt [hunk 14]
```diff
@@ -82,6 +83,73 @@ class ExpressionHandler(frontend: PythonLanguageFrontend) :
...
+    /**
+     * Translates a Python
+     * [`GeneratorExp`](https://docs.python.org/3/library/ast.html#ast.GeneratorExp) into a
+     * [CollectionComprehension].
+     */
+    private fun handleGeneratorExp(node: Python.AST.GeneratorExp): CollectionComprehension {
+        return newCollectionComprehension(node).apply {
+            this.statement = handle(node.elt)
+            this.comprehensionExpressions += node.generators.map { handleComprehension(it) }
+        }
```
[the remark this answers is not in the document]
Comment (KuechA) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt:115]: I have no idea. Is that a type in python??

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt [hunk 15]
```diff
@@ -147,15 +215,23 @@ class ExpressionHandler(frontend: PythonLanguageFrontend) :
...
-                val nextValue = newBinaryOperator("+")
-                nextValue.rhs = start
-                nextValue.lhs = newVal
-                nextValue
-            }
+            joinListWithBinOp("+", values, node)
+        }
+    }
+
+    private fun joinListWithBinOp(
```
[the remark this answers is not in the document]
Comment (KuechA) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt:222]: Not sure if it's understandable

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/edges/collections/UnwrappedEdgeList.kt [hunk 16]
```diff
@@ -70,7 +70,12 @@ class UnwrappedEdgeList<NodeType : Node, EdgeType : Edge<NodeType>>(
     }
 
     override fun subList(fromIndex: Int, toIndex: Int): MutableList<NodeType> {
-        TODO("Not yet implemented")
+        return if (list.outgoing) {
+            list.subList(fromIndex, toIndex).map { it.end }.toMutableList()
+        } else {
+            @Suppress("UNCHECKED_CAST")
```
[the remark this answers is not in the document]
Comment (KuechA) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/edges/collections/UnwrappedEdgeList.kt:76]: No idea, it doesn't seem to be straightforward

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (KuechA) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/ExpressionHandler.kt:122]: There seem to be different classes which can/should be use to type generators (e.g. the return type of a function used as generator) e.g. `typing.Generator`, `collections.abc.Generator` and also some Iterator classes.

**Pull request merged**